kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/ingress/ProxyIngressLayout.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.kubernetes.operator.ingress;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import io.fabric8.kubernetes.api.model.ContainerPort;
+import io.fabric8.kubernetes.api.model.Service;
+import io.fabric8.kubernetes.api.model.ServiceBuilder;
+
+import io.kroxylicious.kubernetes.api.v1alpha1.KafkaProxy;
+import io.kroxylicious.kubernetes.api.v1alpha1.KafkaProxyIngress;
+import io.kroxylicious.kubernetes.api.v1alpha1.VirtualKafkaCluster;
+import io.kroxylicious.proxy.config.VirtualClusterGateway;
+
+import static io.kroxylicious.kubernetes.operator.ProxyDeployment.PROXY_PORT_START;
+import static io.kroxylicious.kubernetes.operator.ResourcesUtil.name;
+
+public record ProxyIngressLayout(List<VirtualClusterLayout> clusterLayouts) {
+
+    public Optional<VirtualClusterLayout> clusterLayout(VirtualKafkaCluster cluster) {
+        return clusterLayouts.stream()
+                .filter(c -> name(c.cluster).equals(name(cluster)))
+                .findFirst();
+    }
+
+    public record VirtualClusterLayout(VirtualKafkaCluster cluster, List<IngressLayout> ingressLayouts) {
+        public List<VirtualClusterGateway> gateways() {
+            return ingressLayouts.stream().map(ingressLayout -> ingressLayout.ingressInstance.gatewayConfig()).toList();
+        }
+
+        public Stream<Service> services() {
+            return ingressLayouts.stream().flatMap(it -> it.ingressInstance().services()).map(ServiceBuilder::build);
+        }
+
+        public Set<IngressConflictException> ingressExceptions() {
+            return ingressLayouts.stream().flatMap(it -> Stream.ofNullable(it.exception)).collect(Collectors.toSet());
+        }
+    }
+
+    public record IngressLayout(IngressInstance ingressInstance, IngressConflictException exception) {
```

Comment:
Not sure what this is. Can you add javadoc?

_Maybe_ it should be called `PortAllocation` (if that's what it is?)

Reply:
allocating ports is part of it, it's also become responsible for determining whether the Ingress is in conflict with other ingresses and reporting that as part of the layout (or allocation).

Reply:
I've refactored to call it `ProxyIngressModel` to remove the confusing layout language, and added an `IngressAllocator` class that does the port allocation and ingress conflict detection.